### PR TITLE
Add range and ObjectMeta to GetResult (#4352) (#4495)

### LINKED ISSUE #4352: Add Range to GetResult::File
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

See https://github.com/apache/arrow-rs/pull/4351#discussion_r1214627268

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4352 
Relates to #4495

# Rationale for this change
 

Not including the byte range results in unexpected behaviour of `GetResult::bytes`.

Additionally it is beneficial to return the `ObjectMeta` alongside the returned data, as this is effectively free, and can be useful for additional data validation

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -729,54 +719,64 @@ impl GetOptions {
...
 
 /// Result for a get request
+#[derive(Debug)]
+pub struct GetResult {
+    /// The [`GetResultPayload`]
+    pub payload: GetResultPayload,
+    /// The [`ObjectMeta`] for this object
+    pub meta: ObjectMeta,
+    /// The range of bytes returned by this request
+    pub range: Range<usize>,
```
Comment (tustvold) [inline object_store/src/lib.rs:729]: I opted to make this required, as it allows for accurate buffer sizing among other things

Comment (alamb) [reply object_store/src/lib.rs:729]: The alternate would be to make it optional (and read file length directly from the metadata if it was set to `None`)? 

If that is the tradeoff I agree that always including the sometimes redundant `range` is a good choice

File: object_store/src/lib.rs [hunk 2]
```diff
@@ -729,54 +719,64 @@ impl GetOptions {
 }
 
 /// Result for a get request
+#[derive(Debug)]
+pub struct GetResult {
```
Comment (alamb) [inline object_store/src/lib.rs:723]: What is the reason for making the fields in this struct `pub`? If they are all `pub` we can't add fields to `GetResult` in the future (such as optional object store specific metadata, for example) without it being a breaking change. 

What do you think about leave the fields as non `pub` and add accessors / and a 

```rust
fn into_parts(self) -> (GetResultPayload, ObjectMeta) {
...
}
```

🤔

File: object_store/src/lib.rs [hunk 3]
```diff
@@ -797,36 +797,12 @@ impl GetResult {
...
-                                    path: path.clone(),
-                                })?;
-
-                            Ok(Some((buffer.into(), (file, path, read != CHUNK_SIZE))))
-                        })
-                    },
-                )
-                .boxed()
+            GetResultPayload::File(file, path) => {
+                local::chunked_stream(file, path, self.range, 8 * 1024)
```
Comment (alamb) [inline object_store/src/lib.rs:803]: I think keeping the name of `CHUNK_SIZE` for the `8 * 1024` would increase this code's readability

File: object_store/src/local.rs [hunk 4]
```diff
@@ -863,7 +866,56 @@ impl AsyncWrite for LocalUpload {
...
 }
 
-pub(crate) fn read_range(file: &mut File, path: &PathBuf, range: Range<usize>) -> Result<Bytes> {
+pub(crate) fn chunked_stream(
+    mut file: File,
+    path: PathBuf,
+    range: Range<usize>,
+    chunk_size: usize,
+) -> BoxStream<'static, Result<Bytes, super::Error>> {
+    futures::stream::once(async move {
```
Comment (alamb) [inline object_store/src/local.rs:875]: I was wondering about using [tokio::fs](https://docs.rs/tokio/latest/tokio/fs/index.html) but it seems like the warnings on that page are still fairly significant

File: object_store/src/local.rs [hunk 5]
```diff
@@ -863,7 +866,56 @@ impl AsyncWrite for LocalUpload {
...
+            file.seek(SeekFrom::Start(range.start as _))
+                .map_err(|source| Error::Seek {
+                    source,
+                    path: path.clone(),
+                })?;
+            Ok((file, path))
+        })
+        .await?;
+
+        let stream = futures::stream::try_unfold(
```
Comment (alamb) [inline object_store/src/local.rs:886]: Do you know if the object_store tests have coverage for files that are greater than 8KB in size? Aka is this code covered by tests?

File: object_store/src/memory.rs [hunk 6]
```diff
@@ -136,17 +138,28 @@ impl ObjectStore for InMemory {
...
+        let meta = ObjectMeta {
+            location: location.clone(),
+            last_modified,
+            size: data.len(),
+            e_tag: None,
+        };
 
+        let (range, data) = match options.range {
+            Some(range) => {
+                ensure!(range.end <= data.len(), OutOfRangeSnafu);
```
Comment (alamb) [inline object_store/src/memory.rs:150]: it occurs to me that these errors would be improved if the included the ranges and lengths as values. I understand that this PR doesn't change the behavior

**Review by alamb — COMMENTED** [review]
The code in this PR looks very good to me. Thank you @tustvold  

Is this a breaking API change to object_store (as it changes [GetResult](https://docs.rs/object_store/0.6.1/object_store/enum.GetResult.html))

I also didn't see any tests for this new feature -- I think we should add some, both to cover the chunking as I mentioned inline as well as to ensure that we don't accidentally break this API or its implementation during future refactors

File: object_store/src/lib.rs (hunk 2 above)
Comment (tustvold) [reply object_store/src/lib.rs:723]: The problem is the various implementations need to be able to construct this, and so this just seemed simpler

File: object_store/src/local.rs (hunk 4 above)
Comment (tustvold) [reply object_store/src/local.rs:875]: Yeah, at least currently tokio::fs has pretty terrible performance charateristics, I would not recommend using it for anything really. Perhaps at some point io_uring will get sufficiently stable, but that will be Linux specific

File: object_store/src/memory.rs [hunk 7]
```diff
@@ -43,11 +45,13 @@ enum Error {
     #[snafu(display("No data in memory found. Location: {path}"))]
     NoDataInMemory { path: String },
 
-    #[snafu(display("Out of range"))]
-    OutOfRange,
+    #[snafu(display(
+        "Requested range {}..{} is out of bounds for object with length {}", range.start, range.end, len
+    ))]
+    OutOfRange { range: Range<usize>, len: usize },
```
Comment (alamb) [inline object_store/src/memory.rs:51]: ❤️

**Review by alamb — APPROVED** [review]
LGTM -- thank you @tustvold

**Pull request merged**